controllers/flinkcluster_reconciler.go

```diff
@@ -573,8 +575,49 @@ func (reconciler *ClusterReconciler) shouldTakeSavepoint(
 // Checks whether it is possible to take savepoint.
 func (reconciler *ClusterReconciler) canTakeSavepoint() bool {
 	var jobSpec = reconciler.observed.cluster.Spec.Job
+	var savepointStatus = reconciler.observed.cluster.Status.Savepoint
+	var jobStatus = reconciler.observed.cluster.Status.Components.Job
 	return jobSpec != nil && jobSpec.SavepointsDir != nil &&
-		!reconciler.isJobStopped()
+		!isJobStopped(jobStatus) &&
+		(savepointStatus == nil ||
+			(savepointStatus.State != SavepointStateProgressing) ||
+			(savepointStatus.State == SavepointStateProgressing && savepointStatus.TriggerID == ""))
+}
+
+// Takes savepoint for a job then update job status with the info.
+func (reconciler *ClusterReconciler) takeSavepointAsync(jobID string) (*v1beta1.SavepointStatus, error) {
+	var log = reconciler.log
+	var cluster = reconciler.observed.cluster
+	var apiBaseURL = getFlinkAPIBaseURL(reconciler.observed.cluster)
+	var triggerSuccess bool
+	var triggerID string
+
+	log.Info("Trigger savepoint.", "jobID", jobID)
+	var retryCount = 0
+	var retriable = func(err error) bool {
+		if httpErr, ok := err.(*flinkclient.HTTPError); ok && httpErr.StatusCode >= 500 {
+			log.Info("Savepoint request is failed", "http error", httpErr, "jobID", jobID, "retry", retryCount)
+			retryCount++
```

Comment:
The same question as above.

Reply:
In my opinion, the possible ways to handle savepoint failures are:

1. Support retry for savepoint creation failure
2. Retry only for savepoint trigger failure
(2-1) backoff retry within iteration (* current implementation)
(2-2) Retry in next iteration and adding triggerRetryCount to CRD
3. No retry

As for the first, I thought it would be better not to retry if the creation of a triggered savepoint ends up failing. Because I thought it would be better to help the user solve the problem by passing the cause of the creation failure to the user faster than delaying the time with multiple retry. And it was thought that the savepoint need not be handled like the Flink cluster components, where the failure state has to be restored in order to maintain the desired state.

However, when the savepoint trigger failed and the failure is retriable, the backoff retry was applied in the iteration. The reason is that in order to handle trigger failures in the next iteration, fields like triggerRetryCount must be added to status, but rather I thought it would be better to apply a few seconds backoff retry and keep the CRD's field simple.

I would like to know what your opinion about these.

Reply:
Sounds good in general, my only concern is that there is no guarantee retry will eventually be successful, in that case, we still fall back to reconcile loop.

Reply:
I agree. I am  going to remove retry here.

Reply:
Fixed